### ISSUE TITLE
Peer review by John Torres

### ISSUE THREAD

**Issue by john2408** [BASTIANGROE/project-bg055-nba#1]
### Peer review team

- Peer review by John Torres
- Names of team members that participated in this review: John Torres

### Goal of the project

The goal of the project is to identify which traditional performance metrics have the strongest correlation with advanced performance metrics (VI, ORtg, and DRtg) in the NBA. The project aims to determine which traditional metrics are already covered by the advanced metrics in the dataset and which additional metrics should be included to better cover the advanced performance metrics.

### Data

- Describe the data used or collected, if any. Is the data adequate for the project?

The data used in the project is collected from various sources, primarily from the official NBA statistics and the website "nbastuffer." The dataset includes:

- **General Information**: Player-specific data such as name, team, position, and age.
- **Traditional Metrics**: Game statistics collected during matches, including GP (Games Played), MPG (Minutes Per Game), USG% (Usage Percentage), FTA (Free Throw Attempts), 2PA (Two-Point Attempts), 3PA (Three-Point Attempts), PPG (Points Per Game), RPG (Rebounds Per Game), APG (Assists Per Game), SPG (Steals Per Game), BPG (Blocks Per Game), and TPG (Turnovers Per Game).
- **Advanced Metrics**: Calculated statistics such as TO% (Turnover Percentage), FT% (Free Throw Percentage), 2P% (Two-Point Percentage), eFG% (Effective Field Goal Percentage), TS% (True Shooting Percentage), P+R (Points + Rebounds), P+A (Points + Assists), P+R+A (Points + Rebounds + Assists), VI (Versatility Index), ORtg (Offensive Rating), and DRtg (Defensive Rating).
The data appears to be adequate for the project as it includes a comprehensive set of both traditional and advanced performance metrics, which are necessary to analyze the correlation between them

### Approach, tools and methods

- Describe the approaches, tools, and methods that will be used.

**Data Collection and Preparation**
- **Tools**: Pandas for data manipulation and storage, NumPy for numerical operations.
- **Methods**: Importing data from CSV files, cleaning and preprocessing the data to ensure it is suitable for analysis. Adtional robust outlier detection using IQR method. 

**Exploratory Data Analysis (EDA)**
- **Tools**: Pandas, Matplotlib, Seaborn for data visualization.
- **Methods**: Descriptive statistics, correlation matrices, and visualizations to understand the relationships between traditional and advanced metrics.

**Feature Selection**
- **Tools**: Scikit-learn for machine learning and statistical analysis.
- **Methods**: Correlation analysis for feature selection and **forward selection method** for selecting the best model using the $R^2$ as metric. Taking as basis a simple linear regression model. 

**Modeling and Analysis**
- **Tools**: Scikit-learn for building and evaluating models.
- **Methods**: Regression analysis to determine the impact of traditional metrics on advanced metrics. Again model selection using the **forward selection method.**. Using both linear and lasso regression model. 

**Visualization and Reporting**
- **Tools**: Matplotlib, Seaborn for creating visualizations; Jupyter Notebooks for interactive analysis and reporting.
- **Methods**: Creating plots and charts to illustrate findings, summarizing results in a clear and concise manner.

### Lack of clarity

- Is there anything that is unclear from the proposal? 

**Modelling Flow**: First linear regression is trained. Then linear regression with Stadardard Scaler. Then Lasso regression. And then the results for linear regression vs lasso regression are presented. This is confusing, it would be better to do first linear and lasso regression without StandardScaler and then present the results. After that, make linear and lasso regression with Scaler and present the results.

### Possible improvements

Provide constructive feedback on how the team might be able to improve their project.Make sure your feedback includes at least one comment on the modeling aspect of the project, but do feel free to comment on aspects beyond the modeling.

- **Better Presentation of Model Comparison**: A bar plot could be used to represent side by side the performance of the models for the response variables in each iterations. This will make the reader easiert to understand how the models improve with the different approaches. 

- **Renaming response variables**: Since 3 response variables are being modelled. it would be better to renamed it to more meaningful names so it is easier to interpret the results, e.g. (VI --> versatility_index, ORtg --> offensive_rating, DRtg --> defensive_rating).

- **Use Multiout Regression**: Since the target is a vector of three response variables *LinearRegression* natively supports multi output regression, so you can predict the three variables in one single model, for example: 

```python
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Sample dataset with 3 response variables
np.random.seed(42)
X = np.random.rand(100, 5)  # 100 samples, 5 features
y = np.random.rand(100, 3)  # 100 samples, 3 response variables

# Split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the Linear Regression model
linear_regressor = LinearRegression()

# Train the model
linear_regressor.fit(X_train, y_train)

# Make predictions
y_pred = linear_regressor.predict(X_test)

- **Feature Selection Methods**: Even though using the correlation metrics to select the best features is a good starting approach it has many downsides: (1) Only Measures Linear Relationships, (2) Ignores Interactions Between Features (3) Sensitive to Outliers. Therefore using `Permutation Importance` or `Mean Decrease Impurity` are good methods to make a more robust feature selection. 

# Evaluate the model
mse = mean_squared_error(y_test, y_pred, multioutput='raw_values')  # MSE for each response variable
print("Mean Squared Error for each output:", mse)

# Overall mean squared error (average across all outputs)
overall_mse = mean_squared_error(y_test, y_pred, multioutput='uniform_average')
print("Overall Mean Squared Error:", overall_mse)
```

### Presentation

What aspect of this project are you most interested in and would like to see highlighted in the final presentation?

- **Model's Performance Presentation**: It would be interested to see an improved result presentation of the model performance. 

- **Additional Feature Selection Methods**: Try using `Permutation importance` or `Mean Decrease Impurity` and compare the model performance afterwards. 

### Organization

- Provide constructive feedback on any issues with file and/or code organization.

- **Functions**: Create some python function to make it easier to iterate through the different model combinations ( e.g. with and withoud Standard Scaler). This will make the code cleaner and easier to read. 

### Further comments

- (Optional) Any further comments or feedback?

In general the project is quite interesting and has a lot of potential to explore. Since you have a lot of predictor variables, you could also benefit a lot from using other models like Random Forest, XGboost or LightGBM, which are quite powerful when in comes to regression prediction, and are proven to be more accurate than linear regression. BoostedTrees also support Multioutput regression.
